### PR TITLE
Add monitoring with Kamon (disabled by default)

### PR DESCRIPTION
**Pull request by pm47**
For now:
- we only track some tasks (especially in the router, but not even
`node_announcement` and `channel_update`)
- all db calls are monitored

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/PimpKamon.scala [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      f
+    } finally {
+      timer.stop()
+    }
+  }
+
+  def timeFuture[T](name: String)(f: => Future[T])(implicit ec: ExecutionContext): Future[T] = {
+    val timer = Kamon.timer(name).withoutTags().start()
+    val res = f
+    res onComplete { case _ => timer.stop }
```
Comment (araspitzu) [inline eclair-core/src/main/scala/fr/acinq/eclair/PimpKamon.scala]: Shouldn't this be `res onComplete { case Success(_) => timer.stop }`? I'd assume we don't want to mix measurements for failed calls with the successful ones.

Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/PimpKamon.scala]: We need to stop the timer somehow? I don't think it matters anyway, because Kamon has first-class handling of distribution of latencies, and we would detect outliers easily.

**Comment by pm47** [comment]
I'm marking this as reviewable to minimize conflicts with other branches. It is just a first step.

Thanks @ivantopo for all the support!

File: eclair-core/src/main/scala/fr/acinq/eclair/PimpKamon.scala [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package fr.acinq.eclair
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/PimpKamon.scala]: nit: no copyright header?

File: eclair-node/src/main/scala/fr/acinq/eclair/Boot.scala [hunk 3]
```diff
@@ -40,6 +42,9 @@ object Boot extends App with Logging {
     implicit val system: ActorSystem = ActorSystem("eclair-node")
     implicit val ec: ExecutionContext = system.dispatcher
     val setup = new Setup(datadir)
+
+    Kamon.init(setup.appConfig)
```
Comment (pm47) [inline eclair-node/src/main/scala/fr/acinq/eclair/Boot.scala]:
```suggestion
    // TODO: disable Kamon for now
    //Kamon.init(setup.appConfig)
```

File: eclair-core/src/main/scala/fr/acinq/eclair/PimpKamon.scala (hunk 1 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/PimpKamon.scala]: I think that's where we'd use tags or some equivalent. 
Usually we should split our measurements on values of a specific tag with Success, Failure1, Failure2, etc depending on our set of possible failures.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala [hunk 4]
```diff
@@ -149,26 +150,36 @@ class ExtendedBitcoinClient(val rpcClient: BitcoinJsonRPCClient) {
...
-      unspent <- isTransactionOutputSpendable(txid, outputIndex, includeMempool = true)
-      fundingTxStatus <- if (unspent) {
-        Future.successful(UtxoStatus.Unspent)
-      } else {
-        // if this returns true, it means that the spending tx is *not* in the blockchain
-        isTransactionOutputSpendable(txid, outputIndex, includeMempool = false).map {
-          case res => UtxoStatus.Spent(spendingTxConfirmed = !res)
+      val span = Kamon.spanBuilder("validate-bitcoin-client").start()
+      for {
+        _ <- Future.successful(0)
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: Why is that needed?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala [hunk 5]
```diff
@@ -149,26 +150,36 @@ class ExtendedBitcoinClient(val rpcClient: BitcoinJsonRPCClient) {
...
-      }
-      tx <- getRawTransaction(txid)
-      unspent <- isTransactionOutputSpendable(txid, outputIndex, includeMempool = true)
-      fundingTxStatus <- if (unspent) {
-        Future.successful(UtxoStatus.Unspent)
-      } else {
-        // if this returns true, it means that the spending tx is *not* in the blockchain
-        isTransactionOutputSpendable(txid, outputIndex, includeMempool = false).map {
-          case res => UtxoStatus.Spent(spendingTxConfirmed = !res)
+      val span = Kamon.spanBuilder("validate-bitcoin-client").start()
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: What's kamon recommendation for naming spans? Do they automatically included the containing namespace/class/object? It's usually recommended to use some kind of namespacing in spans, so either Kamon does it automatically for us or maybe we should do something like `blockchain.bitcoind.rpc.extendedbitcoinclient.validate`?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala [hunk 6]
```diff
@@ -149,26 +150,36 @@ class ExtendedBitcoinClient(val rpcClient: BitcoinJsonRPCClient) {
...
+      for {
+        _ <- Future.successful(0)
+        span0 = Kamon.spanBuilder("getblockhash").start()
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: Is that automagically included as a child span of the `span` defined line 153?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala [hunk 7]
```diff
@@ -149,26 +150,36 @@ class ExtendedBitcoinClient(val rpcClient: BitcoinJsonRPCClient) {
...
+        txid: String <- rpcClient.invoke("getblock", blockHash).map {
+          case json => Try {
+            val JArray(txs) = json \ "tx"
+            txs(txIndex).extract[String]
+          } getOrElse ByteVector32.Zeroes.toHex
         }
-      }
-    } yield ValidateResult(c, Right((Transaction.read(tx), fundingTxStatus)))
+        _ = span1.finish()
+        span2 = Kamon.spanBuilder("getrawtx").start()
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: It looks like we'd benefit a lot from a utility function in KamonExt to wrap a future in a span.
It would be nice to just write `tx <- KamonExt.wrapInSpan("getrawtx")(getRawTransaction(txid))` which would automatically do the `span.finish()`.
The name `wrapInSpan` isn't very good, needs some bikeshedding.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Authenticator.scala [hunk 8]
```diff
@@ -43,6 +44,7 @@ class Authenticator(nodeParams: NodeParams) extends Actor with DiagnosticActorLo
   def ready(switchboard: ActorRef, authenticating: Map[ActorRef, PendingAuth]): Receive = {
     case pending@PendingAuth(connection, remoteNodeId_opt, address, _) =>
       log.debug(s"authenticating connection to ${address.getHostString}:${address.getPort} (pending=${authenticating.size} handlers=${context.children.size})")
+      Kamon.counter("peers.connecting.count").withTag("state", "authenticating").increment()
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Authenticator.scala]: It might be useful to centralize counters for each component (like the `Metrics` object you created for `Peer`).
For example in Authenticator's companion object, define `private val connectingPeersCounter = Kamon.counter("peers.connecting.count")`.
This is usually easier for future maintenance.

File: eclair-core/src/main/scala/fr/acinq/eclair/io/Server.scala [hunk 9]
```diff
@@ -52,6 +53,7 @@ class Server(nodeParams: NodeParams, authenticator: ActorRef, address: InetSocke
   def listening(listener: ActorRef): Receive = {
     case Connected(remote, _) =>
       log.info(s"connected to $remote")
+      Kamon.counter("peers.connecting.count").withTag("state", "connected").increment()
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/io/Server.scala]: This is the same counter used in `Authenticator`. So definitely worth centralizing somewhere to make sure these two actors keep using the same counter correctly

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala [hunk 10]
```diff
@@ -520,7 +536,11 @@ class Router(val nodeParams: NodeParams, watcher: ActorRef, initialized: Option[
         stay
       } else {
         log.info("validating shortChannelId={}", c.shortChannelId)
-        watcher ! ValidateRequest(c)
+        Kamon.runWithContextEntry(shortChannelIdKey, c.shortChannelId) {
+          Kamon.runWithSpan(Kamon.spanBuilder("validate-channel").tag("shortChannelId", c.shortChannelId.toString).start(), finishSpan = false) {
```
Comment (t-bast) [inline eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: How will that span be stopped? I'm not sure it's useful to start it here since we're just sending a message to another actor, it's more useful to do all the spanning inside the watcher isn't it?

File: eclair-node/pom.xml [hunk 11]
```diff
@@ -102,6 +120,32 @@
...
+            <groupId>io.kamon</groupId>
+            <artifactId>kamon-jdbc_${scala.version.short}</artifactId>
+            <version>${kamon.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.kamon</groupId>
+            <artifactId>kamon-annotation_${scala.version.short}</artifactId>
+            <version>2.0.0-RC1</version>
+        </dependency>
+        <!--dependency>
```
Comment (t-bast) [inline eclair-node/pom.xml]: Maybe remove entirely? I don't expect it to be useful to monitor akka-http since we only use it for the RPC API...

File: eclair-node/src/main/scala/fr/acinq/eclair/Boot.scala [hunk 12]
```diff
@@ -40,6 +42,10 @@ object Boot extends App with Logging {
     implicit val system: ActorSystem = ActorSystem("eclair-node")
     implicit val ec: ExecutionContext = system.dispatcher
     val setup = new Setup(datadir)
+
+    // TODO: disable Kamon for now
+    //Kamon.init(setup.appConfig)
```
Comment (t-bast) [inline eclair-node/src/main/scala/fr/acinq/eclair/Boot.scala]: Might be worth making this configurable?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala (hunk 4 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: It's just a limitation of the `for` construct.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala (hunk 6 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: Yes

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala (hunk 4 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: Right, got it.

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala (hunk 5 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/blockchain/bitcoind/rpc/ExtendedBitcoinClient.scala]: That's a good point, let's address this in a 2nd iteration

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 10 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: It is stopped in the `ValidateResult` handler. The point is to be able to measure time spent in the queue, which I think is important to measure backlogs.

File: eclair-node/pom.xml (hunk 11 above)
Comment (pm47) [reply eclair-node/pom.xml]: I think monitoring the API is still useful, and I'm waiting for a compatible version.

File: eclair-node/src/main/scala/fr/acinq/eclair/Boot.scala (hunk 12 above)
Comment (pm47) [reply eclair-node/src/main/scala/fr/acinq/eclair/Boot.scala]: Yes, I was just being lazy

**Comment by pm47** [comment]
@t-bast I addressed the most pressing points. I think we can take care of the rest in a later iteration?

File: eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala (hunk 10 above)
Comment (t-bast) [reply eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala]: Interesting, got it.

File: eclair-node/pom.xml [hunk 13]
```diff
@@ -102,6 +120,32 @@
...
+            <groupId>io.kamon</groupId>
+            <artifactId>kamon-system-metrics_${scala.version.short}</artifactId>
+            <version>${kamon.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.kamon</groupId>
+            <artifactId>kamon-jdbc_${scala.version.short}</artifactId>
+            <version>${kamon.version}</version>
+        </dependency>
+        <dependency>
```
Comment (pm47) [inline eclair-node/pom.xml]:
```suggestion
        <!--dependency>
```

File: eclair-node/pom.xml [hunk 14]
```diff
@@ -102,6 +120,32 @@
...
+            <groupId>io.kamon</groupId>
+            <artifactId>kamon-annotation_${scala.version.short}</artifactId>
+            <version>2.0.0-RC1</version>
+        </dependency>
```
Comment (pm47) [inline eclair-node/pom.xml]:
```suggestion
        </dependency-->
```
